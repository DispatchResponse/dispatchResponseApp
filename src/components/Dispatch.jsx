import React from 'react';
import { Route, NavLink } from 'react-router-dom';
import styled from 'styled-components';
import Map2D from './Map2D';
import Map3D from './Map3D';
import UserSettings from './UserSettings';


export default class Dispatch extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      destinationCoords: null,
      userCoords: null,
    };
    this.getDestinationData = this.getDestinationData.bind(this);
    this.getCurrentLocation = this.getCurrentLocation.bind(this);
  }

  componentDidMount() {
    this.getDestinationData(this.props.dispatchData);
    this.getCurrentLocation();
  }

  getCurrentLocation() {
    navigator.geolocation.getCurrentPosition(position => {
      let userCoords = {
        userLat: position.coords.latitude,
        userLng: position.coords.longitude
      }
      this.setState({userCoords: userCoords})
    }, () => {
      console.log('🐸 ...(denied)');
    });
  }

  getDestinationData(dispatchData) {
    let destinationCoords = {
      destinationLat: parseFloat(dispatchData.latitude),
      destinationLng: parseFloat(dispatchData.longitude)
    }

    this.setState({destinationCoords: destinationCoords})
  }

  render() {

    const alarmColor = this.props.dispatchData.call_category.indexOf('MINOR') > -1
                    || this.props.dispatchData.call_category.indexOf('BOX') > -1 ? 'green'
                     : 'firebrick';


    const DispatchContainer = styled.div`
      display: grid;
      grid-template-columns: 1fr;
    `;

    const MenuItem = styled.div`
        grid-area: 1/2/1/4;
        padding: 0 20px 5% 0;
        z-index: 5;
        text-align: right;
        color: black;
        transform: translateY(-100%);
        font-family: 'Podkova';
        font-size: 2em;
        &:hover{
          text-decoration: underline;
          div {
            display: block;
          }
        }

    `;

    const MenuItems = styled.div`
      display: none;
      position: absolute;
      right: 0;
      text-align: left;
      list-style-type: none;
      background-color: white;
      min-width: 160px;
      box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
      padding: 12px 16px;
      z-index: 5;
      padding: 20px;
      &:hover{
        cursor: pointer;
      }
      li {
        margin: 10px 0 10px 0;
        border-bottom: 1px solid white;
        a {
          text-decoration: none;
          color: black;
          &:hover{
            color: firebrick;
            border-bottom: 1px solid firebrick;
          }
        }
      }
    `;

    const Title = styled.div`
      padding: 20px 0 20px 0;
      display: grid;
      grid-template-rows: 2fr 1fr;
      grid-template-columns: 1fr 5fr 1fr;
      grid-template-areas: '.. description ..'
                           '.. timeout     ..';
      color: white;
      text-align: center;
      background-color: ${alarmColor};
        @media screen and (min-width: 1024px){
          border-radius: 15px 15px 0 0;
        }
    `;

    const Description = styled.div`
      grid-area: description;
      font-size: 3.5em;
      font-family: 'Podkova';
    `;

    const Timeout = styled.div`
      grid-area: timeout;
      font-size: 2.5em;
      font-family: 'Anonymous Pro';
      letter-spacing: 5px;
    `;

    const DispatchDetails = styled.ul`
      padding: 0;
      list-style: none;
      font-family: 'Anonymous Pro';

      li:nth-child(odd) {
        color: firebrick;
        box-shadow: 0 4px 2px -2px lightgray;
        background-color: white;
        border-top: 2px solid white;
        border-bottom: 2px solid firebrick;
        padding: 5px 0 5px 10px;
        font-size: 1.5em;
      }

      li:nth-child(even) {
        color: black;
        font-weight: bolder;
        background-color: white;
        padding: 10px 0 0 10px;
        margin-bottom: 2%;
        font-size: 1.75em;

      }
    `;

    const ApparatusContainer = styled.li`
      display: flex;
      justify-content: space-between;
    `;

    const Apparatus = styled.div`
      color: white;
      text-align: center;
      background-color: black;
      width: 10%;
      font-size: 1.25em;
      border-radius: 50%;
      margin: 2%;
      padding: 2% 2% 2% 3%;
      letter-spacing: 5px;
    `;




    return (

        <DispatchContainer>

        <Title>
          <MenuItem>Menu
            <MenuItems>
              <li>
                <NavLink to="/dispatch-history"> Dispatch History </NavLink>
              </li>
              <li>User Settings</li>
              <li>Snooze Notifications: ON</li>
              {
                this.props.isAdmin
                ? <li>Admin</li>
                : null
              }
            </MenuItems>
          </MenuItem>
          <Description>{this.props.dispatchData.call_category}</Description>
          <Timeout>{this.props.dispatchData.timeout}</Timeout>
        </Title>

        <DispatchDetails>
          <li>Apparatus Assigned</li>
          <ApparatusContainer>
            {
              this.props.dispatchData.assignment.split(', ').map( (apparatus) => {
                return <Apparatus key={apparatus}>{apparatus}</Apparatus>
              })
            }
          </ApparatusContainer>
          <li>Description</li>
          <li>{this.props.dispatchData.call_description}</li>
          <li>Address</li>
          <li>{this.props.dispatchData.location + ", " + this.props.dispatchData.city}</li>
          <li>Nearest Cross Streets</li>
          <li>{ this.props.dispatchData.cross_street }</li>
          <li>Radio Channel & Map Reference</li>
          <li>{ this.props.dispatchData.radio_freq } &nbsp; { this.props.dispatchData.map_ref }</li>
          <li>Dispatch Timeout</li>
          <li>{ this.props.dispatchData.timeout }</li>
          <li>Misc. Details</li>
          <li>{this.props.dispatchData.cfs_remark}</li>
          <li>Navigation</li>
        </DispatchDetails>

        {
        !this.state.userCoords || !this.state.destinationCoords ?  null :
        <Map2D
          userCoords={this.state.userCoords}
          destinationCoords={this.state.destinationCoords}/>
        }

        <DispatchDetails>
          <li>Destination</li>
        </DispatchDetails>

        {
        !this.state.destinationCoords ? null :
        <Map3D destinationCoords={this.state.destinationCoords}/>
        }


        </DispatchContainer>

    )

  }
}
