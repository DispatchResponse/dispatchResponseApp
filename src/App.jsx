import React from 'react';
import { Route } from 'react-router-dom';
import axios from 'axios';
import styled from 'styled-components';
import getCoordinates from './utils/getCoordinates'
import Dispatch from './components/Dispatch';
import DispatchHistory from './components/DispatchHistory';
import UserSettings from './components/UserSettings';
import Admin from './components/Admin';


export default class App extends React.Component {
  constructor(props) {
    super(props);


    this.state = {
      allApparatus: null,
      allCarriers: null,
      dispatchData: null,
      dispatchHistory: null,
      userInfo: null,
      userTracking: null,
      userNotificationStatus: null,
      userApparatusAssignment: null,
      userIsAdmin: false,
      userID: 2,
    };

    this.setAppState = this.setAppState.bind(this);
    this.modifyNotificationStatus = this.modifyNotificationStatus.bind(this);
    this.modifyApparatusAssignment = this.modifyApparatusAssignment.bind(this);
    this.buildApparatusAssigment = this.buildApparatusAssigment.bind(this);
  }

  async componentDidMount() {

    //NOTE:get URL Params, and extract slug and userID.
    //for now slug and userID is hard coded.

    let slug = 'mg08p5p';
    let { userID } = this.state;

    //get Current Dispatch
    await axios.get(`/api/${slug}/${userID}`).then((resp) => {
      console.log("//get Current Dispatch", resp)
      this.setAppState(resp.data[0], 'dispatch');
    })

    //get Dispatch History
    await axios.get('/api/calls').then((resp) => {
      console.log("//get Dispatch History",resp)
      this.setAppState(resp.data, 'dispatchHistory');
    })

    //get All Station Apparatus
    await axios.get('/api/apparatus').then((resp) => {
      console.log("//get All Station Apparatus",resp)
      this.setAppState(resp.data, 'apparatus');
    })

    // get All Carriers
    await axios.get('/api/carriers').then((resp) => {
      console.log("// get All Carriers",resp)
      this.setAppState(resp.data, 'carrier');

    })

    //get User Info
    await axios.get(`/api/users/${userID}`).then((resp) => {
      console.log("//get User Info",resp)
      this.setAppState(resp.data, 'userInfo');
      this.setAppState(resp.data['is_admin'], 'userIsAdmin');
      this.setAppState(resp.data['is_sleeping'], 'userNotificationStatus');

    })

    //get User Tracking
    await axios.get(`/api/tracks/${userID}`).then((resp) => {
      console.log("//get User Tracking",resp)
      this.setAppState(resp.data, 'userTracking');
    })

    this.buildApparatusAssigment()
  }

  setAppState(data, type){
    //setState for dispatch data. used in callback from componentDidMount
      if ( type === 'dispatch') {
      this.setState({dispatchData: data});
    } else if ( type === 'dispatchHistory') {
      this.setState({dispatchHistory: data});
    } else if ( type === 'apparatus' ) {
      this.setState({allApparatus: data});
    } else if ( type === 'carrier' ) {
      this.setState({allCarriers: data});
    } else if ( type === 'userInfo' ) {
      this.setState({userInfo: data});
    } else if ( type === 'userTracking' ) {
      this.setState({userTracking: data});
    } else if ( type === 'userNotificationStatus' ) {
      this.setState({userNotificationStatus: !data});
    } else if ( type === 'userIsAdmin' ) {
      this.setState({userIsAdmin: data});
    }

    return
  }

  buildApparatusAssigment() {
    let userApparatusAssignment = this.state.allApparatus.map( app => {
      for (let i = 0; i < this.state.userTracking.length; i++) {
        if( this.state.userTracking[i]['apparatus_id'] === app['apparatus_id'] ) {
          return {id: app['apparatus_id'], active: true}
        }
      }
      return {id: app['apparatus_id'], active: false}
    })

    this.setState({userApparatusAssignment: userApparatusAssignment})
  }

  modifyNotificationStatus() {
    this.setState({userNotificationStatus: !this.state.userNotificationStatus})
  }

  async modifyApparatusAssignment(e) {
    console.log('modify is running')
    let { userApparatusAssignment, userID } = this.state;

    let appID = e.target.id.split('-').pop();

    let newApparatusAssignment = userApparatusAssignment.map(appItem => {
      if (appItem.id === appID) {
        return {id: appID, active: !appItem.active}
      } else {
        return appItem
      }
    })

    console.log("newApparatusAssignment")
    console.log(newApparatusAssignment)

    let oldAssignmentToDelete = userApparatusAssignment.reduce((acc, item, idx) => {
      if (idx + 1 === userApparatusAssignment.length && item.active) {
        return `${acc + item.id}`
      } else if (idx + 1 === userApparatusAssignment.length) {
        return acc.slice(0, acc.length - 1)
      } else if (item.active) {
        return `${acc + item.id}&`
      } else {
        return acc
      }
    }, '')

    console.log('oldAssignmentToDelete')
    console.log(oldAssignmentToDelete)

    let newAssignmentToAdd = newApparatusAssignment.reduce((acc, item, idx) => {
      if (idx + 1 === newApparatusAssignment.length) {
        return `${acc + item.id}`
      } else if (item.active) {
        return `${acc + item.id}&`
      } else {
        return acc
      }
    }, '')

    console.log('newAssignmentToAdd')
    console.log(newAssignmentToAdd)

    this.setState({userApparatusAssignment: newApparatusAssignment})




    await axios.delete(`/api/tracks/${userID}/${oldAssignmentToDelete}`)
    await axios.post(`/api/tracks/${userID}/${newAssignmentToAdd}`)
    await axios.get(`/api/tracks/${userID}`).then((resp) => {
      console.log("🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 🙏🏾 ")
      console.log(resp)
      this.setAppState(resp.data, 'userTracking');
    })

  }

  render() {

    const AppContainer = styled.div`
        margin -10px;
        /* #### Mobile Phones Portrait #### */
        @media screen and (max-device-width: 480px) and (orientation: portrait){
          background-color: white;
          width: 100vw;
          height: auto;

        }
        /* #### Tablets Portrait or Landscape #### */
        @media screen and (min-device-width: 768px) and (max-device-width: 1024px){
          background-color: white;
          width: 100vw;
          height: auto;
        }

        /* #### Desktops #### */
        @media screen and (min-width: 1024px){
          background-color: white;
          width: 100vw;
          height: auto;
          max-width: 1200px;
          margin: auto;
          margin-top: 10%;
          box-shadow: -3px -3px .7em darkgrey, 3px 3px .7em darkgrey;
          border-radius: 15px;
        }
    `;

    return (

      <div>


        { !this.state.dispatchData ? null : (
          <AppContainer>

           <Route
             exact path="/"
             render={ routeProps =>
               <Dispatch {...routeProps}
                 dispatchData={this.state.dispatchData}
                 notificationStatus={this.state.userNotificationStatus}
                 modifyNotificationStatus={this.modifyNotificationStatus}
                 isAdmin={this.state.userIsAdmin}
               /> }
           />

           <Route
             exact path="/dispatch-history"
             render={ routeProps =>
               <DispatchHistory {...routeProps}
                 dispatchHistory={this.state.dispatchHistory}
                 notificationStatus={this.state.userNotificationStatus}
                 modifyNotificationStatus={this.modifyNotificationStatus}
                 isAdmin={this.state.userIsAdmin}
               /> }
           />

           <Route
             exact path="/user-settings"
             render={ routeProps =>
               <UserSettings {...routeProps}
                 allCarriers={this.state.allCarriers}
                 userInfo={this.state.userInfo}
                 userApparatusAssignment={this.state.userApparatusAssignment}
                 notificationStatus={this.state.userNotificationStatus}
                 modifyNotificationStatus={this.modifyNotificationStatus}
                 modifyApparatusAssignment={this.modifyApparatusAssignment}
                 isAdmin={this.state.userIsAdmin}
               /> }
           />

           <Route
             exact path="/admin"
             render={ routeProps =>
               <Admin {...routeProps}
                 notificationStatus={this.state.notificationStatus}
                 modifyNotificationStatus={this.modifyNotificationStatus}
                 isAdmin={this.state.userIsAdmin}
               /> }
           />

         </AppContainer>

        )}

     </div>



    )
  }
}
