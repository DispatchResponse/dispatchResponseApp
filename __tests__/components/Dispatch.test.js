/*
 * Dispatch.test.js
 */

import React from 'react'
import ReactDOM from 'react-dom'
import Dispatch from '../../src/components/Dispatch'
import App from '../../src/App'
import renderer from 'react-test-renderer'
import toJSON from 'enzyme-to-json'
import { shallow } from 'enzyme'
import { createRenderer } from 'react-test-renderer/shallow'

global.requestAnimationFrme = function(callback) {
  setTimeout(callback, 0)
}

// describe('The Dispatch screen', () => {

  // it('renders a Dispatch component', () => {
  //   const alarmColor = 'firebrick'
  //   const dispatchRenderer = createRenderer();
  //   dispatchRenderer.render(<Dispatch/>);
  //   let dis = dispatchRenderer.getRenderOutput();
  //   expect(dis.type).toBe('div');
  // });

  // it('renders Dispatch component', () => {
  //   const div = document.createElement('div');
  //   ReactDOM.render(
  //     <div>
  //       <Dispatch />
  //     </div>
  //   )
  // })

// })


describe('The App screen', () => {

  it('should match its empty snapshot', () => {
    const tree = renderer.create(
      <App />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
