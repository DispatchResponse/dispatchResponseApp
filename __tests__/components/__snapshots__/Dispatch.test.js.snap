// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The App screen should match its empty snapshot 1`] = `<div />`;
